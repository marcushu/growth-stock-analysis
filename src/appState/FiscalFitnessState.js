class FFstate {

    values = {
        score: 0,

        salesGrowth: 0,
        TTMgrossMargin: 0,
        YAgrossMargin: 0,
        NIgrowth: 0,
        ttmEBITA: 0,
        assetGrowth: 0,
        TLgrowth: 0,
        currentRatio: 0,
        YAcurrentRatio: .0001,
        SharesOut: 0,
        LYsharesOut: 0,
        TLtoEBITDA: 0,
        TTMcf: 0,

        salesGrowthL: 0,
        salesGrowthR: 0,
        TTMgrossMarginL: 0,
        TTMgrossMarginR: 0,
        YAgrossMarginL: 0,
        YAgrossMarginR: 0,
        NIgrowthL: 0,
        NIgrowthR: 0,
        assetGrowthL: 0,
        assetGrowthR: 0,
        TLgrowthL: 0,
        TLgrowthR: 0,
        currentRatioL: 0,
        currentRatioR: 0,
        YAcurrentRatioL: 0,
        YAcurrentRatioR: 0,
        TLtoEBITDAL: 0,
        TLtoEBITDAR: 0
    }

    calculateScore() {
        this.values.score = 0;

        if(this.values.NIgrowth > 0) this.values.score += 1;
        if(this.values.TTMcf > 0) this.values.score += 1;
        if(this.values.NIgrowth > this.values.assetGrowth) this.values.score += 1;
        if(this.values.TTMcf > this.values.NIgrowth) this.values.score += 1;
        if(this.values.NIgrowth > this.values.assetGrowth) this.values.score +=1;
        if(this.values.currentRatio >= this.values.YAcurrentRatio) this.values.score += 1;
        let sharesOut = this.values.LYsharesOut;
        if(this.values.SharesOut < (sharesOut + (sharesOut * (sharesOut * .02)))) this.values.score += 1;
        if(this.values.TTMgrossMargin > this.values.YAgrossMargin) this.values.score += 1;
        if(this.values.salesGrowth > this.values.assetGrowth) this.values.score += 1;
        if(this.values.TLtoEBITDA > 0 && this.values.TLtoEBITDA < 5) this.values.score += 1;
        if(this.values.TLtoEBITDA > 8) this.values.score -= 1;
        if(this.valid({ lval: this.values.TLgrowth, rval: this.TTMcf })) {
            if(this.TLgrowth / this.values.TTMcf < 4) this.values.score += 1;
        }
        

    }

    valid(valuesObject) {
        const l = valuesObject.lval;
        const r = valuesObject.rval;

        return ((!isNaN(l) && !isNaN(r)) && (r > 0 && l > 0));
    }

    setFieldState(valuesObject) {
        this.values[valuesObject.id + 'L'] = +valuesObject.lval;
        this.values[valuesObject.id + 'R'] = +valuesObject.rval;
    }

    divide(valuesObject) {
        if (this.valid(valuesObject)) {
            this.setFieldState(valuesObject);
            this.values[valuesObject.id] = valuesObject.lval / valuesObject.rval;
            this.calculateScore();
        }
    }

    yearOverYear(valuesObject) {
        if (this.valid(valuesObject)) {
            this.setFieldState(valuesObject);
            this.values[valuesObject.id] = ((valuesObject.lval / valuesObject.rval) - 1) * 100;
            this.calculateScore();
        };
    }

    percent = valuesObject => {
        if (this.valid(valuesObject)) {
            this.setFieldState(valuesObject);
            this.values[valuesObject.id] = (valuesObject.lval / valuesObject.rval) * 100;
            this.calculateScore();
        }
    }

    setval = valuesObject => {
        this.values[valuesObject.id] = +valuesObject.val;
        this.calculateScore();
    }
}

export default FFstate