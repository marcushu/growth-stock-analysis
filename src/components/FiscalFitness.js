import React from 'react';
import PropTypes from 'prop-types';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

import FiscalFitnessRow from './FiscalFitnessRow';
import FiscalFitnessSingleField from './FiscalFitnessSingleField';

class FiscalFitness extends React.Component {
  constructor(props) {
    super(props);

    this.extState = this.props.fState;

    this.values = this.extState.values;

    FiscalFitness.propTypes = {
      fState: PropTypes.object,
      submitScore: PropTypes.func
    }

  }

  divide = valuesObject => {
    this.extState.divide(valuesObject);

    this.props.submitScore(this.extState);
  }

  yearOverYear = valuesObject => {
    this.extState.yearOverYear(valuesObject);

    this.props.submitScore(this.extState);
  }

  percent = valuesObject => {
    this.extState.percent(valuesObject);

    this.props.submitScore(this.extState);
  }

  updateField = valuesObject => {
    this.extState.setval(valuesObject);
    
    this.props.submitScore(this.extState);
  }

  render() {
    return (
      <div className="shadow p-4 m-1" style={{ backgroundColor: "rgb(255, 255, 255)" }}>
        <Row>
          <Col>

            <h5>Income Statement</h5>
            <hr style={{ border: "3px solid rgb(86, 141, 212)" }} />
            <FiscalFitnessRow
              scoresId="salesGrowth"
              title="Sales Growth"
              lLabel="TTM Sales"
              rLabel="Year ago TTM Sales"
              lval={this.values.salesGrowthL}
              rval={this.values.salesGrowthR}
              calcFunc={this.yearOverYear} />

            <FiscalFitnessRow
              scoresId="TTMgrossMargin"
              title="TTM Gross Margin"
              lLabel="TTM Gross Profit"
              rLabel="Sales"
              lval={this.values.TTMgrossMarginL}
              rval={this.values.TTMgrossMarginR}
              calcFunc={this.percent} />

            <FiscalFitnessRow
              scoresId="YAgrossMargin"
              title="Year ago Gross Margin"
              lLabel="Year ago gross profit"
              rLabel="Year ago Sales"
              lval={this.values.YAgrossMarginL}
              rval={this.values.YAgrossMarginR}
              calcFunc={this.percent} />

            <FiscalFitnessRow
              scoresId="NIgrowth"
              title="Net income growth"
              lLabel="TTM Net Income"
              rLabel="Year Ago NI"
              lval={this.values.NIgrowthL}
              rval={this.values.NIgrowthR}
              calcFunc={this.yearOverYear} />


            <FiscalFitnessSingleField
              title="TTM EBITA"
              scoresId="ttmEBITA"
              val={this.values.ttmEBITA}
              updateField={this.updateField} />

            <h5>Balance Sheet</h5>
            <hr style={{ border: "3px solid rgb(86, 141, 212)" }} />
            <FiscalFitnessRow
              scoresId="assetGrowth"
              title="Asset Growth"
              lLabel="Total Assets"
              rLabel="Year ago Total Assets"
              lval={this.values.assetGrowthL}
              rval={this.values.assetGrowthR}
              calcFunc={this.yearOverYear} />

            <FiscalFitnessRow
              scoresId="TLgrowth"
              title="Total Liabilities Growth"
              lLabel="Latest TL"
              rLabel="Year ago TL"
              lval={this.values.TLgrowthL}
              rval={this.values.TLgrowthR}
              calcFunc={this.yearOverYear} />

            <FiscalFitnessRow
              scoresId="currentRatio"
              title="Current Ratio"
              lLabel="Latest Current Assets"
              rLabel="Current Liabilities"
              lval={this.values.currentRatioL}
              rval={this.values.currentRatioR}
              calcFunc={this.divide} />

            <FiscalFitnessRow
              scoresId="YAcurrentRatio"
              title="Current Ration(Year Ago)"
              lLabel="YA Current Assets"
              rLabel="YA Current Liabilities"
              lval={this.values.YAcurrentRatioL}
              rval={this.values.YAcurrentRatioR}
              calcFunc={this.divide} />

            <FiscalFitnessRow
              scoresId="TLtoEBITDA"
              title="Total Liabilities to EBITDA"
              lLabel="Total Liabilities"
              rLabel="Year Ago"
              lval={this.values.TLtoEBITDAL}
              rval={this.values.TLtoEBITDAR}
              calcFunc={this.divide} />

            <FiscalFitnessSingleField
              title="TTM Operating Cash Flow"
              scoresId="TTMcf"
              val={this.values.TTMcf}
              updateField={this.updateField} />

            <FiscalFitnessSingleField
              title="Shares Out"
              scoresId="SharesOut"
              val={this.values.SharesOut}
              updateField={this.updateField} />

            <FiscalFitnessSingleField
              title="LY Shares Out"
              scoresId="LYsharesOut"
              val={this.values.LYsharesOut}
              updateField={this.updateField} />

          </Col>
        </Row>
      </div>
    )
  }
}

export default FiscalFitness;