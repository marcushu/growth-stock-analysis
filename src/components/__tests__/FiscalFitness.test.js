import React from 'react';
import { render, cleanup } from '@testing-library/react';
import FiscalFitness from '../FiscalFitness';
import FiscalFitnessState from '../../appState/FiscalFitnessState'

afterEach(cleanup);

test('renders Fiscal Fitness form', () => {
    let ffState = new FiscalFitnessState();

    const { asFragment } = render(
        <FiscalFitness
        submitScore={() => {}}
        fState={ffState} />
    );

    expect(asFragment()).toMatchSnapshot();
});