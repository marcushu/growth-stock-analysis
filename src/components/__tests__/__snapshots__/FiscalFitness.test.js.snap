// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Fiscal Fitness form 1`] = `
<DocumentFragment>
  <div
    class="shadow p-4 m-1"
    style="background-color: rgb(255, 255, 255);"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h5>
          Income Statement
        </h5>
        <hr
          style="border: 3px solid rgb(86, 141, 212);"
        />
        <div
          class="row"
        >
          <div
            class="col-sm-4"
          >
            <br />
            Sales Growth
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  TTM Sales
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Year ago TTM Sales
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-sm-4"
          >
            <br />
            TTM Gross Margin
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  TTM Gross Profit
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Sales
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-sm-4"
          >
            <br />
            Year ago Gross Margin
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Year ago gross profit
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Year ago Sales
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-sm-4"
          >
            <br />
            Net income growth
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  TTM Net Income
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Year Ago NI
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          />
          <div
            class="col"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                >
                  TTM EBITA
                </label>
                <input
                  class="form-control"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col"
          />
        </div>
        <h5>
          Balance Sheet
        </h5>
        <hr
          style="border: 3px solid rgb(86, 141, 212);"
        />
        <div
          class="row"
        >
          <div
            class="col-sm-4"
          >
            <br />
            Asset Growth
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Total Assets
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Year ago Total Assets
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-sm-4"
          >
            <br />
            Total Liabilities Growth
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Latest TL
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Year ago TL
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-sm-4"
          >
            <br />
            Current Ratio
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Latest Current Assets
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Current Liabilities
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-sm-4"
          >
            <br />
            Current Ration(Year Ago)
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  YA Current Assets
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  YA Current Liabilities
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-sm-4"
          >
            <br />
            Total Liabilities to EBITDA
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Total Liabilities
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col-sm-4"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                  for="formGroupEmail"
                >
                  Year Ago
                </label>
                <input
                  class="form-control"
                  id="formGroupEmail"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          />
          <div
            class="col"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                >
                  TTM Operating Cash Flow
                </label>
                <input
                  class="form-control"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col"
          />
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          />
          <div
            class="col"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                >
                  Shares Out
                </label>
                <input
                  class="form-control"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col"
          />
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          />
          <div
            class="col"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="font-weight-light form-label"
                >
                  LY Shares Out
                </label>
                <input
                  class="form-control"
                  style="width: 150px; margin-right: 20px; border: 0px; background-color: rgb(227, 227, 227);"
                  type="text"
                  value="0"
                />
              </div>
            </form>
          </div>
          <div
            class="col"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
